cmd/podman/images/save.go

```diff
@@ -91,11 +97,46 @@ func save(cmd *cobra.Command, args []string) error {
 		return errors.Errorf("--compress can only be set when --format is either 'oci-dir' or 'docker-dir'")
 	}
 	if len(saveOpts.Output) == 0 {
-		fi := os.Stdout
-		if terminal.IsTerminal(int(fi.Fd())) {
+		pipeFile := "saveio"
+		pipeDir, err := ioutil.TempDir(os.TempDir(), "pipeDir")
+		if err != nil {
+			return err
+		}
+		saveOpts.Output = filepath.Join(pipeDir, pipeFile)
+		err = syscall.Mkfifo(saveOpts.Output, 0600)
+		if err != nil {
+			return errors.Wrapf(err, "error creating named pipe")
+		}
+		defer os.Remove(saveOpts.Output)
+		errc := make(chan error)
+		fpipe, err := os.OpenFile(saveOpts.Output, os.O_RDWR, os.ModeNamedPipe)
+		defer fpipe.Close()
+		if err != nil {
+			return err
+		}
+		if terminal.IsTerminal(int(fpipe.Fd())) {
 			return errors.Errorf("refusing to save to terminal. Use -o flag or redirect")
 		}
-		saveOpts.Output = "/dev/stdout"
+		go func() {
+			for {
+				reader := bufio.NewReader(fpipe)
+				_, err = io.Copy(os.Stdout, reader)
+				if err != nil {
+					errc <- err
+					logrus.Errorf("error copying data to /dev/stdout, %v", err)
+					break
+				}
+			}
+			errc <- err
+		}()
+		select {
+		case err := <-errc:
```

Comment:
This will about never trigger here AFAICS; this should happen _after_ sending all the output, and be blocking (modulo timeouts/cancellation) at that point (but it can only be blocking if `fpipe` is closed before, to trigger EOF in the goroutine).

Reply:
The `io.Copy` does not return EOF error. How can I trigger that?

Reply:
Pipes return EOF when there are no descriptors open for writing.

Reply:
I’m sorry, I have finally tried this myself, and it turns out I have missed an important point: `open(2)` blocks _either_ the reader or writer until both are present. So we don’t need (and in fact can’t create) the long-term writer descriptor. OTOH that makes the code even simpler:


```go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"syscall"
)

func fatal(err error) {
	fmt.Fprintf(os.Stderr, "Error: %v", err)
	os.Exit(1)
}

func pipeCopyGoroutine(pipePath string, errc chan<- error) {
	fmt.Fprintf(os.Stderr, "||Opening %s for reading ...\n", pipePath)
	pipeReader, err := os.Open(pipePath)
	if err != nil {
		fatal(err)
	}
	fmt.Fprintf(os.Stderr, "||Read open, copying...\n")
	n, err := io.Copy(os.Stdout, pipeReader)
	fmt.Fprintf(os.Stderr, "||Copied (%v, %v), closing reader... \n", n, err)
	pipeReader.Close()
	fmt.Fprintf(os.Stderr, "||Sending status...\n")
	errc <- err
	fmt.Fprintf(os.Stderr, "||Status sent, goroutine done\n")
}

func setupPipe() (string, func() error) {
	pipeDir, err := ioutil.TempDir(os.TempDir(), "pipeDir")
	if err != nil {
		fatal(err)
	}
	pipePath := filepath.Join(pipeDir, "saveio")
	err = syscall.Mkfifo(pipePath, 0600)
	if err != nil {
		fatal(err)
	}
	fmt.Fprintf(os.Stderr, "Created %s\n", pipePath)

	errc := make(chan error)
	go pipeCopyGoroutine(pipePath, errc)
	return pipePath, func() error {
		fmt.Fprintf(os.Stderr, "Waiting for goroutine...\n")
		err := <-errc
		fmt.Fprintf(os.Stderr, "Goroutine returned %v\n", err)
		os.RemoveAll(pipeDir)
		return err
	}
}

func main() {
//	contents := []byte("Hello pipe\n")
	contentsLen := 1024*1024
	contents := make([]byte, contentsLen, contentsLen)
	for i := 0; i < contentsLen; i+= 1 { contents[i] = 'X' }
	contents[contentsLen-1] = '\n'


	pipePath, cleanup := setupPipe()
	fmt.Fprintf(os.Stderr, "Writing to pipe ...\n")
	ioutil.WriteFile(pipePath, contents, 0600)
	fmt.Fprintf(os.Stderr, "Wrote to pipe, cleaning up...\n")
	err := cleanup()
	fmt.Fprintf(os.Stderr, "Done, error %v\n", err)
}
```
(the debug logging is probably excessive; OTOH do note that this is missing almost all the necessary error handling).

E.g.:
```console
$ go run f.go | sed 's/^XX.*XX$/XX...XX/'
Created /tmp/pipeDir917705942/saveio
Writing to pipe ...
||Opening /tmp/pipeDir917705942/saveio for reading ...
||Read open, copying...
Wrote to pipe, cleaning up...
Waiting for goroutine...
||Copied (1048576, <nil>), closing reader... 
||Sending status...
||Status sent, goroutine done
Goroutine returned <nil>
Done, error <nil>
XX...XX
$
$ go run f.go > /dev/full
Created /tmp/pipeDir330418467/saveio
Writing to pipe ...
||Opening /tmp/pipeDir330418467/saveio for reading ...
||Read open, copying...
||Copied (0, write /dev/stdout: no space left on device), closing reader... 
||Sending status...
Wrote to pipe, cleaning up...
Waiting for goroutine...
Goroutine returned write /dev/stdout: no space left on device
Done, error write /dev/stdout: no space left on device
$
$ go run f.go | (read -rN 1; exec <&-; sleep 5) | sed 's/^XX.*XX$/XX...XX/'
Created /tmp/pipeDir557014252/saveio
Writing to pipe ...
||Opening /tmp/pipeDir557014252/saveio for reading ...
||Read open, copying...
signal: broken pipe
$
$ go run f.go | (sleep 5; cat) | wc
Created /tmp/pipeDir869522253/saveio
Writing to pipe ...
||Opening /tmp/pipeDir869522253/saveio for reading ...
||Read open, copying...
Wrote to pipe, cleaning up...
Waiting for goroutine...
||Copied (1048576, <nil>), closing reader... 
||Sending status...
||Status sent, goroutine done
Goroutine returned <nil>
Done, error <nil>
      1       1 1048576
```

Note that the short-read test requires content larger than the kernel’s pipe buffer, i.e. the short “Hello pipe” content is ~always written successfully even if the consumer ignores it.